primitives/runtime/src/offchain/storage_lock.rs

```diff
@@ -0,0 +1,299 @@
+    }
+
+    fn expired(&self, reference: &Self) -> bool {
+        <Self as Lockable>::current() > *reference
+    }
+
+    fn snooze(&self, deadline: &Self) {
+        let remainder: Duration = self.diff(&deadline);
+        // do not snooze the full duration, but instead snooze max 100ms
+        // it might get unlocked in another thread
+        // consider adding some additive jitter here
+        let snooze = core::cmp::min(remainder.millis(), STORAGE_LOCK_PER_CHECK_ITERATION_SNOOZE);
+        sp_io::offchain::sleep_until(self.add(Duration::from_millis(snooze)));
+    }
+}
+
+/// A persisted guard state.
+///
+/// An in DB persistent mutex for multi access items which are modified
+/// i.e. vecs or sets.
+pub struct StorageLock<'a, L>
+where
+    L: Sized + Lockable,
+{
+    // A storage value ref which defines the DB entry representing the lock.
+    value_ref: StorageValueRef<'a>,
+    deadline: L,
+}
+
+impl<'a, L> StorageLock<'a, L>
+where
+    L: Lockable,
+{
+	/// Create a new storage lock with [default expiry duration](Self::STORAGE_LOCK_DEFAULT_EXPIRY_DURATION_MS).
+    pub fn new<'k>(key: &'k [u8]) -> Self
+    where
+        'k: 'a,
+    {
+        Self {
+            value_ref: StorageValueRef::<'a>::persistent(key),
+            deadline: L::deadline(),
+        }
+    }
+
+    #[inline]
+    fn try_lock_inner(&mut self, new_deadline: L) -> Result<(), Option<L>> {
+        let now = L::current();
+        let res = self
+            .value_ref
+            .mutate(|s: Option<Option<L>>| -> Result<L, Option<L>> {
+                match s {
+                    // no lock set, we can safely acquire it
+                    None => Ok(new_deadline),
+                    // lock is set, but it's old. We can re-acquire it.
+                    Some(Some(deadline)) if now.expired(&deadline) => Ok(new_deadline),
+                    // lock is present and is still active
+                    Some(Some(deadline)) => Err(Some(deadline)),
+                    _ => Err(None),
+                }
+            });
+        match res {
+            Ok(Ok(_)) => Ok(()),
+            Ok(Err(_deadline)) => Err(None),
+            Err(e) => Err(e),
+        }
+    }
+
+
+	/// Attempt to lock the storage entry.
+	///
+	/// Returns a lock guard on success, otherwise an error containing `None` in
+	/// case the mutex was already unlocked before, or if the lock is still held
+	/// by another process `Err(Some(expiration_timestamp))`.
+    pub fn try_lock<'b>(&'b mut self) -> Result<StorageLockGuard<'a, 'b, L>, Option<L>>
+    where
+        'a: 'b,
+    {
+        match self.try_lock_inner(self.deadline) {
+            Ok(_) => Ok(StorageLockGuard::<'a, 'b> { lock: Some(self) }),
+            Err(e) => Err(e),
+        }
+    }
+
+	/// Try grabbing the lock until its expiry is reached.
+	///
+	/// Returns an error if the lock expired before it could be caught.
+    pub fn spin_lock<'b, 'c>(&'b mut self) -> StorageLockGuard<'a, 'b, L>
```

Comment:
Why not use proper context switching locks instead of spin locks? Is deadline guaranteed to be short?

Reply:
No, the deadline is really a recovery mechanism in case of a crash / missing unlock. PM

Reply:
@arkpar not sure if this is worth the expansion of this PR or it could be done in a follow up which then introduces `fn lock(..)` using conditionals?

Reply:
Since this now "snooze" when contended "spin_lock" would be incorrect naming. As it implies that the functions spins in a busy loop.
Also, I still don't understand why this whole thing can't be replaced by a `parking_lot::Mutex` which supports fair locking and deadlines. It seems it would simplify code by a lot.

Reply:
> Since this now "snooze" when contended "spin_lock" would be incorrect naming. As it implies that the functions spins in a busy loop.

It does, but with some extra sleep to prevent consuming 100% of CPU. What name would you suggest then? Just `lock`?

> Also, I still don't understand why this whole thing can't be replaced by a parking_lot::Mutex which supports fair locking and deadlines. It seems it would simplify code by a lot.

There are two use cases that we are addressing with `StorageLock`:
1. Synchronizing storage access in an (unlikely) event of multiple workers trying to modify the same entries. (for instance: We don't want multiple workers to waste time pruning the same session index from a map).
2. Grace period after doing some heavy computations and sending a transaction to be included on-chain. (case for `Guard::forget`; for instance: when we submit offchain phragmen election we don't want to send another one right away, but rather we give the transaction few blocks to be included on-chain).

The first case can indeed be solved by adding extra host functions, that would allow them to acquire a unique namespaced lock (`parking_lot::Mutex`).
In the second case we would rather keep the grace period even if the node is restarted, so the lock is better kept in the storage than memory.

The main reason was to just avoid introducing extra host APIs, especially given that multiple OCW competing for the same lock should really be super rare - I was only hoping to enclose in a nicer API what we already do manually in some pallets.
But indeed, unfortunately this evolved into something more complicated and perhaps the Mutex-lock-host-functions would be a better & simpler approach here - I'd let @drahnr decide what works best, but block-based locks (with perhaps limited API) would still be useful to get in since they would simplify some code in offchain phragmen & im-online pallets.

Reply:
Right. Definitely was not suggesting changing the API. I was under the impression that `sleep` may simply be replaced by waiting on a shared mutex instance that would be placed in e.g. `BlockAndTime`. But after discussing it with @drahnr I can see that this is not applicable here indeed. 

> It does, but with some extra sleep to prevent consuming 100% of CPU. What name would you suggest then? Just lock?

Yes. `lock` suggests the the call is blocking already. Spinlocks are usually CPU busy and are intended for something very short term. At least that's what I'd think if I saw the function named like that.

Reply:
I agree renaming `spin_lock` -> `lock` would make sense.